### ISSUE TITLE
Run Schedulers Does Not Create Orders for Components of Components

### ISSUE THREAD

**Issue by matt454357** [odoo/odoo#3626]
Impacted versions:
- 8.0

Steps to reproduce:
- Create stockable products Prt1, Asm2, and Asm3
  - Prt1 = Buy
  - Asm2 = Manufacture
  - Asm3 = Manufacture
- Create BOM
  - Asm3
    - Asm2
      - Prt1
- Create reordering rules IN THIS ORDER
  - Prt1
  - Asm2
  - Asm3
- Confirm sales order for Asm3
- Run Schedulers only once

Current behavior:
- Creates an MO for Asm3

Expected behavior:
- Create an MO for Asm3
- Create an MO for Asm2
- Create an RFQ for Prt1

**Comment by spadae22** [comment]
Is this going to be merged in 8.0?

**Comment by mart-e** [comment]
Hello,

I agree it's not optimal to execute the orders only in the creation order (needs to run several time the scheduler).
However, I am not convinced adding a low level code field is the way to go. This one will apply some complicated logic based on BoM components. The list of components is only one criteria, people may have other ones. It looks over-complex (with the risk of users not understanding what this field mean) to cover only a subset of cases.

Instead, I propose to add a sequence or priority field on the orderpoint. Way simpler and obvious to the user what is the effect.
What do you think?

However, both solutions requires a new field so it can not be integrated into 8.0.
Except if somebody suggest a third solution, we could explain on the procurement form with a message that the orderpoints will be executed in the creation orders.

**Comment by AlanBell** [comment]
If the scheduler keeps a track of what it did, then it could re-run itself if it did anything. It would then repeat until it had nothing remaining to do. Creating a priority on the order points is going to be really hard to manage, especially when you have generic sub-assemblies that might be used in multiple high level BOMs

**Comment by scosist** [comment]
The user never sees the low-level-code (LLC). The user does not create a priority on the order points (reorder rules). The user has no direct control over the LLC. It is not user-editable; it is a hidden field. The LLC (#4315) is calculated at scheduler runtime based on the BOM hierarchy the user has defined.

**Comment by gdgellatly** [comment]
Perhaps a simple way is through the fact that the scheduler typically runs
using a new cursor.

Move the if new cursor statement to the top of the while loop and also add
an else statement for the final commit (i.e create a while else loop)

Now the ids search will pick up the newly created ids from the run and
rerun.

Too simple?  Missed something?

**Comment by gdgellatly** [comment]
Sorry, ignore that, got the issue wrong, is to do with reorder rules.

**Comment by anajuaristi** [comment]
It's not so easy. The problem is that in Odoo, you need to have lauched all levels manufacturing orders to have the procurements created for each level. 
On the other side scheduler is not taking in account the date where you really need the material but is taking global quantity to calculate quantities from min-max rules. 
Third problem is that each procurement is creating one single manufacturing order... so it's complicated you could group several procurements (even those who are generated by scheduler) in a single one as it's done with purchases orders... 

Mixin all this premises is quite complex to manage procurements+ mo generation and control. 

To handle this issues we have created procurement_plan_ " modules set where you could create procurements for all materials included on BOM levels starting from one procurement that includes  a manufacturing product. At the same time it will also create a "stock reserve" for each procurement created so you can easily separate "real" movements from "reserved" movements

Afterwards, we also have the "stock_planning_" modules set where you can see date by date how many products you will need from today to the date you select (day by day, week to week... and so on )

Modules are published in 
https://github.com/odoomrp/odoomrp-wip and 
https://github.com/odoomrp/odoomrp-utils
Sorry for indirect publicity but we have been fighting with this issue during months and maybe the solution we found could be the starting point to solve it definitively  

Thank you:
Ana

**Comment by matt454357** [comment]
Ana, thank you for your comment.  They fix some important issues, but I don't see how your modules fix this issue.

**Comment by matt454357** [comment]
I don't understand the resistance to using Low-Level Code.  It is the industry standard.  It is a fundamental concept in Material Requirements Planning.  Every big-name ERP uses Low-Level Coding (SAP, Oracle, Microsoft Dynamics, etc.).  The implementation is simple and efficient.

**Comment by alaa9jo** [comment]
This issue is still not resolved in v9.0 , is there any workaround ?

**Comment by AlanBell** [comment]
I don't see why getting the scheduler to re-run itself until there is nothing left to do would be a bad approach. That works now with manual re-runs, but it isn't clear how many times you have to run it, you are generally left with running it until the specific thing that you want to happen is achieved, or if you have a maximum BOM depth of 3 then you might always run it 3 times.

**Comment by matt454357** [comment]
@alaa9jo,
Module mrp_llc implements low-level-coding for v9.0.  It hasn't had much testing, but it's not complicated:
https://github.com/asphaltzipper/azi-odoo-modules

Be aware that the stock_calendar addon may not be compatible with mrp_llc.  If you are using stock_calendar, then mrp_llc needs changes to the _procure_orderpoint_confirm() method.

@AlanBell,
Check the discussion on pull request #4315.  Everyone, everywhere, uses low level coding.  It is easy to implement, and very efficient.  Much more efficient than re-running the scheduler.  My deepest BOM has 6 levels.  Run Schedulers completes in 12 minutes.  That makes 72 minutes.  Adding low-level-code to the calculation adds 2 or 3 seconds.  That makes 12:02.

**Comment by AlanBell** [comment]
OK, that looks like it is happening and will work. My impression was that the scheduler was pretty fast if there was nothing for it to do so re-running wouldn't be much of an overhead. Adding the llc might take 2 or 3 seconds, but it would mean the scheduler does more scheduling so your 12 minutes might become 24 minutes if it was only doing half it's job before.

**Comment by lucasczpnk** [comment]
Hello, @matt454357 .
I have a similar need that needs this fix. 
Do you know if this is going to be implemented for v8.0?
Or if there's any other solution or workaround for this issue by now?

Any help is appreciated.
Thanks in advance.

**Comment by matt454357** [comment]
Hi @lucasczpnk,

PR #4315 probably will never be incorporated into Odoo Community Edition, and certainly not v8.0. 
 They have thrown their effort into an Enterprise module called Master Production Schedule (mrp_mps).  Their module works around the low-level-code issue, but still does not provide the desired planning functionality.

You could incorporate low-level-coding yourself, using our pull request.  However, there are other conceptual problems with Odoo's scheduler algorithm that make it impractical.  The whole idea of Run Schedulers and _procure_orderpoint_confirm() is misguided.

The Enterprise Master Production Schedule is very limited.  Odoo SA seems unwilling to implement a real Material Requirements Planning system.  I guess it's simply not a priority for them.

For v10.0, I created my own complete material requirements planning system.  It includes:

 * a master schedule
 * an accurate and manageable scheduling algorithm
     * low-level-code
     * time-phasing of planned orders (weekly buckets)
     * easy plan changes (planned orders can be overwritten/deleted as the master schedule changes)
 * reporting of status to the user (big plans can take several minutes to calculate)
     * progress of the scheduling algorithm
     * missing BOMs
     * missing reorder rules
 * tools for organizing and analyzing planned orders
 * the user converts planned orders to procurement orders when due to start

The modules can be found here: https://github.com/asphaltzipper/azi-odoo-modules.
I have no demo data, documentation or tests.  We are still making improvements, but it does work.  Back-porting to v8.0 would probably require significant work.
The applicable modules are:

 * mrp_master_schedule
 * mrp_llc
 * mrp_mrp
 * simulated_pick

Eventually, I will separate these modules into their own repository, and do the demo data, documentation and tests.

**Comment by Yenthe666** [comment]
Hi @matt454357 ,

Thank you for your bug report.
At this point Odoo 8 is no longer supported as Odoo only supports three versions (9.0, 10.0, 11.0 at the moment).
If you can still reproduce this issue on Odoo 9, Odoo 10 or Odoo 11 we encourage you to create a new bug report.

Regards,
Yenthe

**Comment by Ark74** [comment]
Even though this issue was reported on version 8
I can confirm it's present all the way until version 10.

Somebody else should confirm it on version 11.

**Comment by Yenthe666** [comment]
Hi @Ark74

Please do create a new bug report if you can still reproduce it. This way it'll contain the newest info and we've got a new and clean report to handle.

Regards,
Yenthe
